spec/Spec/BlsEC/FQP.cry

```diff
@@ -0,0 +1,328 @@
+/**
+ * This module provides support for polynomial extension fields.
+ * Specifically, we need to support Fq2 and Fq12.
+ * @see https://github.com/ethereum/py_ecc/blob/70c194edd8d1eb1457805442f0162499dbc0aac5/py_ecc/fields/field_elements.py#L204
+ */
+module Spec::BlsEC::FQP where
+
+import Spec::BlsEC::Field
+import Spec::BlsEC::Poly(poly_extended_euclidean)
+
+/**
+ * A polynomial field extension type contains a list of
+ *  coefficients which are elements of Fq.
+ */
+type FQP n = [n](Z Fq)
+
+/**
+ * A G2 point is comprised of two Fq2 elements, or sets of coefficients, each of degree two.
+ * Each element is comprised of a real and imaginary coeffsicient.
+ */
+type Fq2 = FQP 2
+
+/**
+ * The modulus coefficients for the quadratic extension field Fq2.
+ * @see https://github.com/ethereum/py_ecc/blob/70c194edd8d1eb1457805442f0162499dbc0aac5/py_ecc/fields/field_properties.py#L30
+ */
+FQ2_MODULUS_COEFFICIENTS : Fq2
+FQ2_MODULUS_COEFFICIENTS = [1, 0]
+
+/**
+ * Addition of two sets of coefficients.
+ * @see https://github.com/ethereum/py_ecc/blob/70c194edd8d1eb1457805442f0162499dbc0aac5/py_ecc/fields/field_elements.py#L232
+ * ```repl
+ * :prove G2_y ~+~ G2_x == G2_y_plus_x
+ * ```
+ */
+(~+~) : {n} (fin n, n >= 1) => FQP n -> FQP n -> FQP n
+(~+~) coeffs1 coeffs2 = [c1i + c2i | c1i <- coeffs1 | c2i <- coeffs2]
+
+/**
+ * Subtraction of two sets of coefficients.
+ * @see https://github.com/ethereum/py_ecc/blob/70c194edd8d1eb1457805442f0162499dbc0aac5/py_ecc/fields/field_elements.py#L240
+ * ```repl
+ * :prove G2_y ~-~ G2_x == G2_y_minus_x
+ * ```
+ */
+(~-~) : {n} (fin n, n >= 1) => FQP n -> FQP n -> FQP n
+(~-~) coeffs1 coeffs2 = [c1i - c2i | c1i <- coeffs1 | c2i <- coeffs2]
+
+/**
+ * Multiplication of two sets of coefficients.
+ * @see https://github.com/ethereum/py_ecc/blob/70c194edd8d1eb1457805442f0162499dbc0aac5/py_ecc/fields/field_elements.py#L251
+ * ```repl
+ * :prove G2_x ~*~ G2_x == G2_xSquared
+ * :prove G2_x ~*~ G2_x ~*~ G2_x == G2_xCubed
+ * ```
+ */
+(~*~) : {n} (fin n, n >= 2) => FQP n -> FQP n -> FQP n
+(~*~) coeffs1 coeffs2 = reduce_while b (2 * `(n) - 1) where
+    b = poly_multi coeffs1 coeffs2
+
+    reduce_while : {a, n} (fin n, n >= 2, fin a, a == 2 * n - 1) => [a](Z Fq) -> Integer -> [n](Z Fq)
```

Comment:
I believe Cryptol prefers you to not have the same variable for type parameters in helper functions

Reply:
Even though the two should be the same? That's what I want to indicate with reusing `n`.

Reply:
Oh, yeah that doesn't even work --- apparently I forgot to rerun the tests. Yikes.